src/common/timing/time-keeper.js

```diff
@@ -0,0 +1,90 @@
+import { gosNREUM } from '../window/nreum'
+import { globalScope } from '../constants/runtime'
+
+/**
+ * Class used to adjust the timestamp of harvested data to New Relic server time. This
+ * is done by tracking the performance timings of the RUM call and applying a calculation
+ * to the harvested data event offset time.
+ */
+export class TimeKeeper {
+  /**
+   * Represents the browser origin time corrected to NR server time.
+   * @type {int}
+   */
+  #correctedOriginTime
+
+  /**
+   * Represents the page origin time offset of the NR server time. This is the difference
+   * between the local browser time and the NR server time.
+   * @type {int}
+   */
+  #serverOffset
+
+  static getTimeKeeperByAgentIdentifier (agentIdentifier) {
+    const nr = gosNREUM()
+    return Object.keys(nr?.initializedAgents || {}).indexOf(agentIdentifier) > -1
+      ? nr.initializedAgents[agentIdentifier].timeKeeper
+      : undefined
+  }
+
+  get correctPageOriginTime () {
+    return this.#correctedOriginTime
+  }
+
+  /**
+   * Process a rum request to calculate NR server time.
+   * @param rumRequest {XMLHttpRequest} The xhr for the rum request
+   * @param rumRequestUrl {string} The full url of the rum request
+   */
+  processRumRequest (rumRequest, rumRequestUrl) {
+    const responseDateHeader = rumRequest.getResponseHeader('Date')
+    if (!responseDateHeader) {
+      throw new Error('Missing date header on rum response.')
+    }
+
+    const resourceEntries = globalScope.performance.getEntriesByName(rumRequestUrl, 'resource')
+    if (!Array.isArray((resourceEntries)) || resourceEntries.length === 0) {
+      throw new Error('Missing rum request performance entry.')
+    }
+
+    let medianRumOffset = 0
+    this.#serverOffset = 0
+    if (typeof resourceEntries[0].responseStart === 'number' && resourceEntries[0].responseStart !== 0) {
+      // Cors is enabled and we can make a more accurate calculation of NR server time
```

Comment:
FMI, in what cases does rum service not provide cors header allowing any domain, or at least the [Timing-Allow-Origin](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Timing-Allow-Origin)?

Reply:
The RUM call did not have the Timing-Allow-Origin header or the necessary CORS Expose Header header. I had to add those in a separate PR. That said, we should never actually hit the alternate code as there is no browser we support that does not support this branch of code. If a customer is using a proxy and alters the headers, thats on them. So, I could remove the alternate code path or leave it as a "just in case".